library/ssl_tls13_client.c

```diff
@@ -575,6 +575,42 @@ static int ssl_tls13_parse_cookie_ext( mbedtls_ssl_context *ssl,
     return( 0 );
 }
 
+static int ssl_tls13_write_cookie_ext( mbedtls_ssl_context *ssl,
+                                       unsigned char *buf,
+                                       unsigned char *end,
+                                       size_t *olen )
+{
+    unsigned char *p = buf;
+
+    *olen = 0;
+
+    if( ssl->handshake->verify_cookie == NULL )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 3, ( "no cookie to send; skip extension" ) );
+        return( 0 );
+    }
+
+    MBEDTLS_SSL_DEBUG_BUF( 3, "client hello, cookie",
+                           ssl->handshake->verify_cookie,
+                           ssl->handshake->verify_cookie_len );
+
+    MBEDTLS_SSL_CHK_BUF_PTR( p, end, ssl->handshake->verify_cookie_len + 6 );
+
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello, adding cookie extension" ) );
+
+    MBEDTLS_PUT_UINT16_BE( MBEDTLS_TLS_EXT_COOKIE, p, 0 );
+    MBEDTLS_PUT_UINT16_BE( ssl->handshake->verify_cookie_len + 2, p, 2 );
+    MBEDTLS_PUT_UINT16_BE( (size_t) ssl->handshake->verify_cookie_len, p, 4 );
```

Comment:
We missed that in the HRR PR but the fix is not to cast `ssl->handshake->verify_cookie_len` to `size_t` here but for `verify_cookie_len` to be of type uint16_t.
We should also probably rename the `verify_cookie` and `verify_cookie_len` to `cookie` and `cookie_len` and in ssl_misc.h describe that they are used to store the HelloVerifyRequest cookie (DTLS) or HelloRetryRequest cookie (TLS 1.3).

Reply:
If we change the type of verify_cookie_len to uint16_t, will compiler complain the data loss when assgin the verify_cookie_len to unsigned char pointer, just like the assignment in library/ssl_cli.c, line 930. ```
"            MBEDTLS_SSL_DEBUG_BUF( 3, "client hello, cookie",
                              ssl->handshake->verify_cookie,
                              ssl->handshake->verify_cookie_len );

            ***p++ = ssl->handshake->verify_cookie_len;**

            MBEDTLS_SSL_CHK_BUF_PTR( p, end,
                                     ssl->handshake->verify_cookie_len );
            memcpy( p, ssl->handshake->verify_cookie,
                       ssl->handshake->verify_cookie_len );
            p += ssl->handshake->verify_cookie_len;
"
```

Reply:
ok then to avoid ugly casts the best solution is probably to have another "_len" variable: `hrr_cookie_len` ? Thus 3 fields, something like:
```C
unsigned char *cookie;
unsigned char verify_cookie_len;
uint16_t hrr_cookie_len;
```

Reply:
Can I find the fix in any exist code or any PR?